### PR TITLE
Update `wasi-crypto`

### PR DESCRIPTION
**Pull request by rjzak**
* This has been discussed in issue #4610.
* Update `wasi-crypto`
* Test Wasm https://github.com/rjzak/wasi-crypto-example/releases/tag/v0.0.1

CC: @npmccallum, @sunfishcode

### CODE REVIEW

File: crates/wasi-crypto/Cargo.toml [hunk 1]
```diff
@@ -13,7 +13,7 @@ edition = "2021"
 
 [dependencies]
 anyhow = "1.0"
-wasi-crypto = { path = "spec/implementations/hostcalls/rust", version = "0.1.5" }
+wasi-crypto = { git = "https://github.com/WebAssembly/wasi-crypto", rev="fa309ac", features = ["rcrypto"], default-features = false }
```
Comment (sunfishcode) [inline crates/wasi-crypto/Cargo.toml:16]: wasmtime-wasi-crypto is published on crates.io, which I believe requires that all dependencies be published on crates.io. Can you publish this revision to a new crates.io release, and then depend on that?

Comment (rjzak) [reply crates/wasi-crypto/Cargo.toml:16]: I'll ask.

Comment (jedisct1) [reply crates/wasi-crypto/Cargo.toml:16]: A new version has been pushed to `crates.io`.

File: crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs [hunk 2]
```diff
@@ -263,8 +265,10 @@ impl From<guest_types::PublickeyEncoding> for PublicKeyEncoding {
             guest_types::PublickeyEncoding::Pkcs8 => PublicKeyEncoding::Pkcs8,
             guest_types::PublickeyEncoding::Pem => PublicKeyEncoding::Pem,
             guest_types::PublickeyEncoding::Sec => PublicKeyEncoding::Sec,
-            guest_types::PublickeyEncoding::CompressedSec => PublicKeyEncoding::CompressedSec,
+            //guest_types::PublickeyEncoding::CompressedSec => PublicKeyEncoding::CompressedSec,
```
Comment (jedisct1) [inline crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs:268]: Leftover from debugging code?

File: crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs [hunk 3]
```diff
@@ -263,8 +265,10 @@ impl From<guest_types::PublickeyEncoding> for PublicKeyEncoding {
...
             guest_types::PublickeyEncoding::Local => PublicKeyEncoding::Local,
+
+            _ => todo!("From guest_type::PublicKeyEncoding to PublicKeyEncoding"),
```
Comment (jedisct1) [inline crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs:271]: raw, pkcs8, pem, sec and local are the only possible values. Why is this `todo!()` required?

File: crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs (hunk 2 above)
Comment (rjzak) [reply crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs]: Removed

File: crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs (hunk 3 above)
Comment (rjzak) [reply crates/wasi-crypto/src/wiggle_interfaces/asymmetric_common.rs]: Removed. Previously I was getting errors about unhandled patterns in the `match` statement, but that seems to be resolved.

**Comment by jedisct1** [comment]
@rjzak -- Looking good now! Thanks!

Would you mind resolving the conflict so it can be merged?

**Comment by rjzak** [comment]
@jedisct1 I'd like to resolve the conflict, but I changed the Cargo.toml file and need it to have my version, not what this thing is that Git seems to be trying to merge. How can I force it to be my version?

**Comment by rjzak** [comment]
@jedisct1 I could use some advice regarding some of the CI errors. A lot of these are not related to wasi-crypto.

**Comment by rjzak** [comment]
I added the nightly requirement for `wasmtime-wasi-crypto` since `wasi-crypto` uses Cargo's "per-package-target" https://github.com/WebAssembly/wasi-crypto/compare/67d9821f2763f4626a97c70c411ccb92a2047712...fa309ace2bb335bb08efc724a18721e8217d5977#diff-9c249d46387da48fcc9d73b78e7a99a00bb825ae6b5fa222b60f6b27f9a4d2a0. Is this ok @sunfishcode?

**Comment by jedisct1** [comment]
@rjzak I don't think `nightly` is required (at least I hope it's not...)

However, the errors about dead code were likely due to the example `main.rs` file in `wasi-crypto/.../bindings/rust/src`. 

This has been fixed upstream, so an update to a more recent commit should fix the issues we are seeing here as well.

**Comment by rjzak** [comment]
Nightly was only needed for running the test case for Wasi-Crypto, not for compiling the Wasmtime-Wasi-Crypto module.

I'll look at updating the Wasi-Crypto submodule.

**Comment by jedisct1** [comment]
Yes, but the test case depends on the Rust example bindings.

**Comment by rjzak** [comment]
I've updated the Wasi-Crypto git submodule, may I have a re-run of CI? @sunfishcode or @jedisct1 
This should fix one of the CI issues about unused import.

**Comment by rjzak** [comment]
Abandoning in favour of future PR against the main branch.

**Pull request closed without merging**